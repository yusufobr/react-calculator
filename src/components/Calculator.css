td {
  text-align: center;
  padding-bottom: 2px;
  padding-right: 2px;
  width: 25%;
}

td:hover {
  background-color: #00000008;
}

input[type="text"] {
  height: 50px;
}

.operator {
  background-color: rgb(255, 146, 82);
}
