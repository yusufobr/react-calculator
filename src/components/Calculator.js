import React from 'react';
import './Calculator.css';

export default function Calculator() {
  return (
    <div className="text-[#333] text-sm font-bold p-12">
      <table className="h-[300px] w-[350px] mx-auto bg-[#E0E0E0]">
        <thead>
          <tr>
            <td colSpan="4">
              <input type="text" value="0" className="w-full pr-1 text-right text-xl font-normal text-white bg-[#858693]" />
            </td>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <input type="button" value="AC" />
            </td>
            <td>
              <input type="button" value="+/-" />
            </td>
            <td>
              <input type="button" value="%" />
            </td>
            <td className="operator">
              <input type="button" value="รท" />
            </td>
          </tr>
          <tr>
            <td>
              <input type="button" value="7" />
            </td>
            <td>
              <input type="button" value="8" />
            </td>
            <td>
              <input type="button" value="9" />
            </td>
            <td className="operator">
              <input type="button" value="x" />
            </td>
          </tr>
          <tr>
            <td>
              <input type="button" value="4" />
            </td>
            <td>
              <input type="button" value="5" />
            </td>
            <td>
              <input type="button" value="6" />
            </td>
            <td className="operator">
              <input type="button" value="-" />
            </td>
          </tr>
          <tr>
            <td>
              <input type="button" value="1" />
            </td>
            <td>
              <input type="button" value="2" />
            </td>
            <td>
              <input type="button" value="3" />
            </td>
            <td className="operator">
              <input type="button" value="+" />
            </td>
          </tr>
          <tr>
            <td colSpan="2">
              <input type="button" value="0" />
            </td>
            <td>
              <input type="button" value="." />
            </td>
            <td className="operator">
              <input type="button" value="=" />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  );
}
