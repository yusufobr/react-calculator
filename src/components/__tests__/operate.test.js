import { fireEvent, render, screen } from '@testing-library/react';
import Calculator from '../Calculator';
import operate from '../../logic/operate';

describe('operate', () => {
  describe('add', () => {
    test('should add two numbers correctly', () => {
      expect(operate(2, 2, '+')).toMatchSnapshot();
      expect(operate(5, 5, '+')).toMatchSnapshot();
    });
  });

  describe('subtract', () => {
    test('should subtract two numbers correctly', () => {
      expect(operate(5, 3, '-')).toMatchSnapshot();
      expect(operate(1, 1, '-')).toMatchSnapshot();
    });
  });

  describe('multiply', () => {
    test('should multiply two numbers correctly', () => {
      expect(operate(5, 3, 'x')).toMatchSnapshot();
      expect(operate(2, 1, 'x')).toMatchSnapshot();
    });
  });

  describe('divide', () => {
    test('should divide two numbers correctly', () => {
      expect(operate(5, 5, 'รท')).toMatchSnapshot();
      expect(operate(10, 5, 'รท')).toMatchSnapshot();
      expect(operate(5, 0, 'รท')).toMatchSnapshot();
    });
  });

  describe('modulo', () => {
    test('should find modulo of two positive numbers correctly', () => {
      expect(operate(10, 3, '%')).toMatchSnapshot();
      expect(operate(15, 7, '%')).toMatchSnapshot();
      expect(operate(5, 0, 'รท')).toEqual("Can't divide by 0.");
    });
  });

  describe('unknown operation', () => {
    test('should throw an error for unknown operation', () => {
      expect(() => operate(5, 3, '$')).toThrow("Unknown operation '$'");
      expect(() => operate(10, 5, '@')).toThrow("Unknown operation '@'");
    });
  });
});

describe('Button component', () => {
  test('should update the state when clicked', () => {
    render(<Calculator />);

    const buttonElements = screen.getAllByRole('button');
    const acButton = buttonElements.find((button) => button.value === 'AC');
    fireEvent.click(acButton);

    const inputElement = screen.getByRole('textbox');
    expect(inputElement).toHaveValue('0');
  });
});
