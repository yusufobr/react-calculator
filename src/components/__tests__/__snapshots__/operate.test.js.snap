// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add adds two numbers correctly 1`] = `"4"`;

exports[`add adds two numbers correctly 2`] = `"10"`;

exports[`divide divide two numbers correctly 1`] = `"1"`;

exports[`divide divide two numbers correctly 2`] = `"2"`;

exports[`divide divide two numbers correctly 3`] = `"Can't divide by 0."`;

exports[`modulo should find modulo of two positive numbers correctly 1`] = `"1"`;

exports[`modulo should find modulo of two positive numbers correctly 2`] = `"1"`;

exports[`multiply multiply two numbers correctly 1`] = `"15"`;

exports[`multiply multiply two numbers correctly 2`] = `"2"`;

exports[`operate add should add two numbers correctly 1`] = `"4"`;

exports[`operate add should add two numbers correctly 2`] = `"10"`;

exports[`operate divide should divide two numbers correctly 1`] = `"1"`;

exports[`operate divide should divide two numbers correctly 2`] = `"2"`;

exports[`operate divide should divide two numbers correctly 3`] = `"Can't divide by 0."`;

exports[`operate modulo should find modulo of two positive numbers correctly 1`] = `"1"`;

exports[`operate modulo should find modulo of two positive numbers correctly 2`] = `"1"`;

exports[`operate multiply should multiply two numbers correctly 1`] = `"15"`;

exports[`operate multiply should multiply two numbers correctly 2`] = `"2"`;

exports[`operate subtract should subtract two numbers correctly 1`] = `"2"`;

exports[`operate subtract should subtract two numbers correctly 2`] = `"0"`;

exports[`subtract subtracts two numbers correctly 1`] = `"2"`;

exports[`subtract subtracts two numbers correctly 2`] = `"0"`;
